### DISCUSSION TITLE
jetstream newbie questions: How do I inspect the contents of a workqueue stream to see what all messages are pending?

### DISCUSSION THREAD

**Discussion by vamshiaruru** [nats-io/nats-server#4778] [Q&A] [answered]
Hi everyone, I am exploring nats to see if we can use it in our product as queue system. I created a stream with **workqueue** retention policy and published a bunch of messages into it. Now, I want to see the messages in the stream that are yet to be consumed. I tried this: (EVENTS is the name of stream)
```
nats stream view EVENTS
``` 
I get the following error: 
```
nats: error: work queue stream contents can not be viewed
```
So how do I view all messages in the stream without consuming them, say for debugging reasons? Do I have to write a consumer that fetches all messages, prints them, but doesn't ack them? Is there a CLI utility for this? In general, is there a UI that helps with management of streams/KV/objects?

Similarly, if I do `nack` with delay of say, 10 days, is there a way to check what all messages are waiting to be delivered again like this? Or how many times a message has been nacked, or how many delivery attempts have been made for this message?

Unrelated question: I need to be able to implement deadletter queues as well, and the official documentation says to listen on `$JS.EVENT.ADVISORY.CONSUMER.MAX_DELIVERIES.<STREAM>.<CONSUMER>` subject. Is there a collection of such recipes for nats that help implementing common queue related stuff like this?

Thanks in advance!

**Comment by ripienaar** [comment]
The latest NATS CLI version supports viewing WQs if you enable direct get option when creating the stream.

**Reply by vamshiaruru** [reply]
Thanks for the response, I upgraded my cli version. Now I get this error:
```
nats stream view EVENTS
nats: error: multiple non-filtered consumers not allowed on workqueue stream (10099)
```

**Reply by ripienaar** [reply]
probably not much I can do about that one for the moment, can fix for a future nats cli release if you want to open a issue in `natscli` project with this

**Reply by vamshiaruru** [reply]
thanks, I will open an issue for this. 

Until a new version of nats-cli is released, the way to inspect is to fetch all messages and not ack them? Is there a way to check how many times the message has been nacked, or when it is expected to be delivered again? If I nack with delay of 10 days, will it still appear when I fetch all messages?

Any tips on the second part of my question?
```
I need to be able to implement deadletter queues as well, and the official documentation says to listen on $JS.EVENT.ADVISORY.CONSUMER.MAX_DELIVERIES.<STREAM>.<CONSUMER> subject. Is there a collection of such recipes for nats that help implementing common queue related stuff like this?
```

**Reply by ripienaar** [reply]
Data such as redelivery etc would be in the consumer. You can not see that by inspecting messages only have access to that when a delivery is done. Same for when next it will be delivered. 

Only place for that kind of hints is docs, natsbyexample.com and the community slack.

**Reply by vamshiaruru** [reply]
So is there a way to get that info from the consumer then? I don't see any methods in the code that I have seen so far.

In general, how would you debug these sort of things, where you need visibility into the message lifetime (as in who all processed this, how many times was this processed, when will it be processed again etc)?

**Reply by ripienaar** [reply]
Answered above. Only available when delivered.

**Reply by vamshiaruru** [reply]
Oops, sorry, missed that. 

I raised an issue in the nats-cli repo here: https://github.com/nats-io/natscli/issues/920 for the first bug.

**Comment by pySilver** [comment]
@vamshiaruru did you manage to find an easy way to view messages in a queue?

**Reply by derekcollison** [reply, marked as answer]
Use nats cli from top of main and stream view or SCP.

**Reply by pySilver** [reply]
@derekcollison it complains about `filtered consumer not unique on workqueue stream` when I invoke `nats stream view`

**Reply by pySilver** [reply]
to be exact: `nats: error: multiple non-filtered consumers not allowed on workqueue stream (10099)`

**Reply by derekcollison** [reply]
You need the cli from top of main which uses direct gets vs consumers.

```
curl -sf https://binaries.nats.dev/nats-io/natscli/nats@main | sh
```

**Reply by dozsa** [reply]
> You need the cli from top of main which uses direct gets vs consumers.
> 
> ```
> curl -sf https://binaries.nats.dev/nats-io/natscli/nats@main | sh
> ```

Noticed this fix is not present in the December 1.6 natscli release. Was wondering how this was not part of the release as it was in main in Oct. Any indication when it will be part of a natscli release? Thanks.

**Reply by ripienaar** [reply]
The December release was a big fix - while supporting WQ was quite a big change and so was left over for the next feature release.

**Comment by pySilver** [comment]
oh! @derekcollison thanks! this one works!
